api/src/main/java/org/apache/iceberg/types/Comparators.java

```diff
@@ -108,6 +109,15 @@ public int compare(StructLike o1, StructLike o2) {
         return 0;
       }
 
+      if (o1 instanceof StructProjection && o2 instanceof StructProjection) {
```

Comment:
This is a pretty big behavior change and I'm not sure it's the right place to fix this issue. The previous behavior created a comparator which only cared about those fields which are passed in the Constructor, this changes that to instead look at the raw lengths of the projections instead.

Can you explain a little more about the bug you are trying to fix and write some test cases which are currently failing to demonstrate the issue?

Reply:
Updated the PR to compare the number of projected fields.

The issue is that 2 projections with null and missing field values are considered equals. That leads to incorrect PartitionMap construction: 

Below PartitionProjections are equal, so in PartitionMap we'll have just 1 random entry:
````
"ice_orc(company_id=null)"
"ice_orc(company_id=null/dept_id=null)"
"ice_orc(company_id=null/dept_id=null/team_id=null)"
````

Reply:
added test: `TestMetadataTableScansWithPartitionEvolution#testPartitionSpecEvolutionNullValues`

Reply:
I think this is going to break other things, because most other places in the code we do want those partitions to be equal

Reply:
let's see, but right now that leads to invalid partition list evaluation

Reply:
an alternative solution could be replacing `NULL` values in a struct with `__HIVE_DEFAULT_PARTITION__`
````
StructLike partitionKey = newPartitionKey(file.partition(), "__HIVE_DEFAULT_PARTITION__");
StructLike partition = PartitionUtil.coercePartition(
    partitionType, table.specs().get(file.specId()), file.partition());
partitions.get(partition).update(file, snapshot);
````
however, I don't think it's a better approach since this hack should be applied in multiple places and could potentially lead to new issues. 

I believe that PartitionProjections containing VOID and NULL fields should not be considered equal. If all tests pass under the assumption that they should be, this assumption may not be accurate.

Reply:
@RussellSpitzer as per your recommendation, created a localized fix for PartitionsTable. Please take a look.